### DISCUSSION TITLE
Feature Request: Negative prompt in commandline

### DISCUSSION THREAD

**Discussion by ClashSAN** [azuritecoin/OnnxDiffusersUI#48] [Ideas]
Could you add the option to add negative prompting in commandline? I would really appreciate this option as using cpu, and the browser + gradio open ends up adding 10 seconds more per iteration. It's claimed that disabling the gradio loading bar improves  inference speed [here.](https://github.com/AUTOMATIC1111/stable-diffusion-webui/wiki/Features#loading) 

I'm might end up trying to install this on a lightweight linux distribution in order to generate a higher maximum size in one go without using swap memory. This kit is awesome by the way, the model conversion just works, it is only 18% slower than bes-dev/openvino , lower sizes are an option (finetuned miniSD), and I am able to use better sampling (DPM++ 2M) thanks to diffusers, in 6-10 steps.

**Comment by azuritecoin** [comment]
If you're going with Linux, you should use ROCm instead of ONNX. You'll get a 3-4x better performance.

Automatic1111's webUI works with AMD GPUs on Linux.

**Reply by ClashSAN** [reply]
I want to try getting sd to run on a portable usb, with non-gpu systems

**Comment by azuritecoin** [comment]
Added negative prompt to txt2img_onnx.py

BTW, if you're going to focus on non-gpu systems, you should take a look at the fp16 version of the model. It works in CPU ONNX, but not using DirectML.

**Reply by ClashSAN** [reply]
I used the ckpt to onnx script and append `--fp16`
currently in similiar situation to: https://github.com/huggingface/diffusers/issues/1083#issuecomment-1312952577


On linux, I can run the fp32 model, I did the fp16 conversion via the script successfully, but it doesn't seem to accept fp16:


> Traceback (most recent call last):
>   File "/home/user/Documents/test/onnxcpu/OnnxDiffusersUI/txt2img_onnx.py", line 99, in <module>
>     images = pipe(
>   File "/home/user/Documents/test/onnxcpu/OnnxDiffusersUI/onnxcpu/lib/python3.9/site-packages/diffusers/pipelines/stable_diffusion/pipeline_onnx_stable_diffusion.py", line 275, in __call__
>     noise_pred = self.unet(sample=latent_model_input, timestep=timestep, encoder_hidden_states=text_embeddings)
>   File "/home/user/Documents/test/onnxcpu/OnnxDiffusersUI/onnxcpu/lib/python3.9/site-packages/diffusers/onnx_utils.py", line 61, in __call__
>     return self.model.run(None, inputs)
>   File "/home/user/Documents/test/onnxcpu/OnnxDiffusersUI/onnxcpu/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 200, in run
>     return self._sess.run(output_names, input_feed, run_options)
> onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Unexpected input data type. Actual: (tensor(float)) , expected: (tensor(float16))

the models are not hardware independent 

>Hardware-independent FP16 checkpoints are still WIP, as they require still-unreleased onnxruntime updates.

**Reply by azuritecoin** [reply]
There's one thing you need to change in the code. In the `get_latents_from_seed` function change:  
`image_latents = rng.standard_normal(latents_shape).astype(np.float32)`  
to  
`image_latents = rng.standard_normal(latents_shape).astype(np.float16)`

**Reply by ClashSAN** [reply]
ty,  following that change, I was able to run fp16, however running in this mode resulted in slower generation times than fp32. `14.6s/it` on a steam deck to `23s/it`, with 512x512.

for 256x256,  It is `2.77s/it` at fp32 and `5s/it` at fp16

Maybe something went wrong during conversion, or it is unoptimized for cpu use.

**Reply by azuritecoin** [reply]
did the memory usage go down at least?

**Reply by ClashSAN** [reply]
it seems to for the lower sized generations

512x512
10.1 for fp32
7.9 but then later jumps and stays at 12.6 fp16

256x256
6.9 for fp32 
5.4 for fp16

**Reply by ClashSAN** [reply]
@azuritecoin I was able to run your txt2img script with a quantized int8 onnx model, it does very well and uses very low memory. At 256x256, it uses 0.9 gb of ram, with 1.3 gb peak. This will be good to use with any model finetuned at 256x256. It takes the same amount of time for inference.

Do you happen to know how to use the onnxruntime-openvino acceleration with your txt2img script? 

from the diffusers package, in line 145 in import_utils.py:

    _onnxruntime_version = "N/A"
    _onnx_available = importlib.util.find_spec("onnxruntime") is not None
    if _onnx_available:
    candidates = ("onnxruntime", "onnxruntime-gpu", "onnxruntime-directml", "onnxruntime-openvino")
    _onnxruntime_version = None

**Reply by azuritecoin** [reply]
I don't know of any development to get diffusers to work with openvino. If I understand correctly, you will need to convert original stable diffusion/diffusers models to openvino models then run the model using ONNX.

If you got stable diffusion running on openvino then it's not that hard to create a UI interface using gradio interface.

**Reply by ClashSAN** [reply]
thanks for the info. I thought the direct-ml and openvino backends can both use regular onnx models, but I guess not. bes dev seemed to have converted the unet in a way that does not allow different sizes. Only 1 person was successful converting a model (wd1.3) for that openvino repo, I tried but ran into issues.
